type_classes.md

```diff
@@ -160,6 +281,11 @@ export Inhabited (default)
 Sometimes we want to think of the default element of a type as being an *arbitrary* element, whose specific value should not play a role in our proofs.
 For that purpose, we can write ``arbitrary`` instead of ``default``. We define ``arbitrary`` as an *opaque* constant.
 Opaque constants are never unfolded by the type checker.
+
+BUGBUG: what does "Opaque constants" mean and what makes it opaque.  Is it the square brackets?
```

Comment:
Sounds like you are saying all constants are opaque by definition?

Reply:
this definition of unfolding is not super clear to me, is there an even simpler way to say it?

Reply:
Yes, the purpose of the `constant` keyword is to make an opaque declaration.

Maybe its more clear with an example:
```lean
namespace Unfold

def id : α → α := λ x => x

example : ∀ (x : α), id x = x := by
  intro x
  simp only [id] -- this is unfolding the definition of id

end Unfold
```
So unfolding the definition is a bit simply said "plug in what's on the right hand side of := in a `def`" (the function also gets directly applied here so we don't really see the definition being pluged in since the goal gets instantly closed).

Now if `id` was a `constant`:
```lean
namespace Unfold

constant id : α → α := λ x => x

example : ∀ (x : α), id x = x := by
  intro x
  simp only [id]

end Unfold
```
The `simp only` line doesn't actually do anything since the definition is not accessible to us as all constants are opaque. The goal is in fact impossible to prove like this.